[modelio] Update for iOS 10.3 beta 1

VincentDondain opened the pull request:
- Added `MDLAssetTest` (copy from `MDLObjectTest`).
- Indexers tests.

spouliot commented on tests/monotouch-test/ModelIO/MDLObject.cs:
```diff
@@ -63,6 +63,19 @@ public void GetBoundingBox ()
+		public void IndexerTest ()
+		{
+			using (var obj = new MDLObject ()) {
+				var key = new Protocol ("MDLComponent");
```
I'm not sure we want to expose `Protocol` like this in the API. At best it's not very type safe, I'd better use

> var key = new Protocol (typeof (MDLComponent));

My first thought was to _abuse_ `[BindAs]` but a conversion from `Protocol` to `System.Type`, with additional checks to ensure it's a protocol, can be costly if done repetitively...  

Maybe we can just provide a `[Wrapper]` that calls a conversion method ? and also expose that as a new `Protocol.ctor(System.Type)` ? 

@rolfbjarne @dalexsoto thoughts ?

rolfbjarne replied:
I think just a `[Wrapper]` that calls a conversion method (or a new Protocol ctor) would be fine.

BTW your example is wrong :smile: It should be:

> var key = new Protocol (typeof (IMDLComponent));

dalexsoto replied:
Yep, agree on just the `Wrapper` 👍

spouliot commented on src/ModelIO/MDLAsset.cs:
```diff
@@ -1,12 +1,23 @@
+				return ObjectForKeyedSubscript (key);
+			}
+			set {
+				SetObject (value, key);
```
that's confusing because your new `SetObject` (even if `internal`) is not symmetric with the existing `GetObject`

spouliot commented on src/modelio.cs:
```diff
@@ -90,6 +90,28 @@ interface MDLAsset : NSCopying
+		[iOS (10,3), TV (10,2), Mac (10,12,4)]
+		[Export ("componentConformingToProtocol:")]
+		[return: NullAllowed]
+		IMDLComponent IsComponentConforming (Protocol protocol);
```
`Is` makes it sound boolean, but this is not what it returns. Seems to me that `Get` is the action (not question) being done.

Also it seems to be related with the previous method, so `GetComponent` would match.

VincentDondain replied:
@spouliot I totally agree with you but I wanted to keep it close to `MDLObject`'s `IsComponentConforming`...

Most of the changes in this PR are about aligning `MDLObject` and `MDLAsset`.

spouliot replied:
Let's fix the bad one with an `[Obsolete]` and add a new, accurate, name :)

rolfbjarne replied:
Maybe `GetConformingComponent`? That's a bit more descriptive and closer to what ObjC calls it (which makes finding it easier).

rolfbjarne replied:
Ah, I see the `SetComponent` method now, `GetComponent` would be the opposite method (I assume at least, since Apple named them quite differently).

VincentDondain replied:
Yea so in the headers `setComponent:forProtocol:` and `componentConformingToProtocol:` have the exact same description:

> Extensible component support that allows user of ModelIO to customize MDLAssets to fit their format and workflow.

So I'm going for `GetComponent` instead of `IsComponentConforming`.

spouliot commented on src/ModelIO/MDLAsset.cs:
```diff
@@ -1,12 +1,23 @@
 		}
+
+		[iOS (10,3), TV (10,2), Mac (10,12,4)]
+		public IMDLComponent this [Protocol key] {
```
there's already a `this` that returns something else... check FxDG but I don't think we should expose it this way. That can get confusing source wise when the argument type is not clear.

VincentDondain replied:
FxDG (Framework Design Guideline)?

spouliot replied:
yup :)

VincentDondain replied:
Mmh here's the guideline: https://msdn.microsoft.com/en-us/library/ms229006(v=vs.110).aspx

> **AVOID** indexers with parameter types other than System.Int32, System.Int64, System.String, System.Object, or an enum.

So `Protocol` is definitely not recommended.

> **DO NOT** provide more than one family of overloaded indexers in one type.

I guess that validates your point @spouliot. As I understand it it's fine to "overload" an indexer (as in have an indexer that's using an other indexer in the same type) but it's wrong to create a 2nd and different family of indexers.

spouliot commented on src/ModelIO/MDLObject.cs:
```diff
@@ -0,0 +1,17 @@
+namespace XamCore.ModelIO {
+	public partial class MDLObject {
+		[iOS (10,3), TV (10,2), Mac (10,12,4)]
+		public IMDLComponent this [Protocol key] {
```
should match how `MDLAsset` turns out...

spouliot commented on src/modelio.cs:
```diff
@@ -90,6 +90,28 @@ interface MDLAsset : NSCopying
+
+		[Internal]
+		[Export ("setObject:forKeyedSubscript:")]
+		void SetObject ([NullAllowed] IMDLComponent obj, Protocol key);
```
are those just doing the same as the `setComponent:forProtocol:` and `componentConformingToProtocol:` above ? in that case maybe we should not expose them ?

VincentDondain replied:
Yes so `objectForKeyedSubscript:`'s description is:

> Allows shorthand [key] syntax for `componentConformingToProtocol:`

@spouliot correct me if I'm wrong but this is basically the indexer I implemented initially (`public IMDLComponent this [Protocol key]` which was sentenced to die because there's already one).

spouliot commented on src/modelio.cs:
```diff
@@ -986,13 +1008,26 @@ interface MDLNormalMapTexture
+
+		[Internal]
+		[Export ("setObject:forKeyedSubscript:")]
+		void SetObject ([NullAllowed] IMDLComponent obj, Protocol key);
```
same (not sure we need those, tests would confirm)

spouliot commented on src/modelio.cs:
```diff
@@ -986,13 +1008,26 @@ interface MDLNormalMapTexture
 
 		[Export ("componentConformingToProtocol:")]
 		[return: NullAllowed]
 		IMDLComponent IsComponentConforming (Protocol protocol);
```
that's old, but badly named :(

VincentDondain replied:
Ah well you found the origin of my `IsComponentConforming`...

spouliot commented on src/modelio.cs:
```diff
@@ -1068,6 +1103,16 @@ interface MDLObjectContainerComponent : MDLComponent, INSFastEnumeration
 		void RemoveObject (MDLObject @object);
 
+		[iOS (10,3), TV (10,2), Mac (10,12,4)]
+		[Abstract]
```
breaking change, you cannot add `abstract` methods as existing subclasses won't have them

VincentDondain replied:
Ah damn, good point I missed that.

spouliot commented on src/modelio.cs:
```diff
@@ -1068,6 +1103,16 @@ interface MDLObjectContainerComponent : MDLComponent, INSFastEnumeration
+		MDLObject GetObjectAtIndexedSubscript (nuint index);
+
+		[iOS (10,3), TV (10,2), Mac (10,12,4)]
+		[Abstract]
```
same

monojenkins commented:
Build failure

VincentDondain commented:
Introspection-mac
```
[FAIL] Selector not found for ModelIO.MDLAsset : objectForKeyedSubscript:
[FAIL] Selector not found for ModelIO.MDLAsset : setObject:forKeyedSubscript:
[FAIL] Selector not found for ModelIO.MDLObject : objectForKeyedSubscript:
[FAIL] Selector not found for ModelIO.MDLObject : setObject:forKeyedSubscript:
```

VincentDondain commented:
Unrelated tests failures (due to the move to Xcode8.3):
apitest - https://bugzilla.xamarin.com/show_bug.cgi?id=51799
monotouch-test - https://bugzilla.xamarin.com/show_bug.cgi?id=51800
mini watchOS - https://bugzilla.xamarin.com/show_bug.cgi?id=50419
don’t link watchOS - https://bugzilla.xamarin.com/show_bug.cgi?id=50419

rolfbjarne commented on src/modelio.cs:
```diff
@@ -1068,6 +1103,16 @@ interface MDLObjectContainerComponent : MDLComponent, INSFastEnumeration
+		[iOS (10,3), TV (10,2), Mac (10,12,4)]
+		[Abstract]
+		[Export ("objectAtIndexedSubscript:")]
+		MDLObject GetObjectAtIndexedSubscript (nuint index);
```
Just `GetObject` would be better.

rolfbjarne commented on tests/monotouch-test/ModelIO/MDLAsset.cs:
```diff
@@ -0,0 +1,76 @@
+		[Test]
+		public void IndexerTest ()
+		{
+			using (var obj = new MDLAsset ()) {
```
I think you need a version check here, for Xcode 8.3.

monojenkins commented:
Build failure

VincentDondain commented:
Alright so:

1. I'm open to suggestions on how to assert that `Type` is a `Protocol` in `ObjCRuntime/Protocol.cs` (not sure at all I'm doing the right thing but it works :P)
2. I removed the Indexers I added for both `MDLAsset` and `MDLObject` because they did not follow the Framework Design Guidelines, therefore I removed `MDLObject.cs` that was new in this PR and the tests I created since they were related to the new manual code.

For `MDLObject`, based on the description, I'm really not sure that `objectAtPath` should become an indexer like `MDLAsset`'s `objectAtIndex`.

The first one being:
> Return the object at the specified path, or nil if none exists there

The second being:

> Returns the top-level object at the specified index in the asset.
> **index**: An index in the asset’s list of top-level objects; between zero and the value of the count property.

monojenkins commented:
Build failure

rolfbjarne commented on src/ObjCRuntime/Protocol.cs:
```diff
@@ -24,6 +24,18 @@ public Protocol (string name)
 
+		public Protocol (Type type)
+		{
+			this.handle = Class.GetHandle (type);
```
This is not correct.

1. First check that the type is an interface, otherwise it can't be a protocol:

        if (!type.IsInterface)
            throw new ArgumentException (string.Format ("'{0}' is not a protocol.", type.FullName));

2. Then get the Protocol attribute, and get the name of the protocol from there:

        var protocols = type.GetCustomAttributes (typeof (ProtocolAttribute), false);
        if (protocols.Count == 0)
            throw ...

        var protocol = protocols [0];
        var protocolName = protocol.Name;
        handle = objc_getProtocol (protocolName);
        if (handle == IntPtr.Zero)
            throw ...

VincentDondain replied:
Mmh didn't realized I could dive into the managed type like that! Much better thanks.

monojenkins commented:
Build failure

spouliot commented on src/ObjCRuntime/Protocol.cs:
```diff
@@ -24,6 +24,23 @@ public Protocol (string name)
 
+		public Protocol (Type type)
+		{
+			var exception = new ArgumentException (string.Format ("'{0}' is an unknown protocol", type.FullName));
```
This creates an exception instance for every protocol created (thru this .ctor). This is **not** efficient, use something like:

```
if (type.IsInterface) {
   foreach (var pa in type.GetCustomAttributes<ProtocolAttribute> (false)) {
      handle = objc_getProtocol (pa.Name);
      if (handle != IntPtr.Zero)
         return;
   }
}
if (handle == IntPtr.Zero)
   new ArgumentException (string.Format ("'{0}' is an unknown protocol", type.FullName));
```

rolfbjarne replied:
Also having different exception messages for each condition gives you valuable information (which check failed?).

spouliot commented on src/modelio.cs:
```diff
@@ -90,6 +90,27 @@ interface MDLAsset : NSCopying
+		[Export ("components", ArgumentSemantic.Copy)]
+		IMDLComponent[] Components { get; }
+
+		[Internal]
```
maybe` [Advanced]` but not `[Internal]`. The main reason is that `Type` is easier (and type safe) to use, but it would be costly in a loop, e.g.

```
foreach (var component in list)
   x.SetComponent (type);
```

vs

```
var p = new Protocol (type);
foreach (var component in list)
   x.SetComponent (p);
```

It also makes it possible to  override the method in a subclass.

spouliot commented on src/modelio.cs:
```diff
@@ -90,6 +90,27 @@ interface MDLAsset : NSCopying
+		[Wrap ("SetComponent (component, new Protocol (type))")]
+		void SetComponent (IMDLComponent component, Type type);
+
+		[Internal]
```
same

spouliot commented on src/modelio.cs:
```diff
@@ -986,13 +1007,31 @@ interface MDLNormalMapTexture
+#if XAMCORE_4_0
+		[Internal]
+#endif
+		[Obsolete ("Use SetComponent (Type protocol)")]
```
remove both `[Obsolete]` and `[Internal]`

spouliot commented on src/modelio.cs:
```diff
@@ -986,13 +1007,31 @@ interface MDLNormalMapTexture
+#if XAMCORE_4_0
+		[Internal]
+#endif
+		[Obsolete ("Use GetComponent (Type protocol)")]
```
Obsolete `IsComponentConforming` but expose `GetComponent (Protocol protocol)` for the same reasons as above

dalexsoto commented on src/ObjCRuntime/Protocol.cs:
```diff
@@ -24,6 +24,19 @@ public Protocol (string name)
 				throw new ArgumentException (String.Format ("'{0}' is an unknown protocol", name));
 		}
 
+		public Protocol (Type type)
```
❤️

monojenkins commented:
Build failure

monojenkins commented:
Build failure

VincentDondain commented:
Unrelated test failures
- https://bugzilla.xamarin.com/show_bug.cgi?id=52162 - monotouch-test - Unified 32bits
- https://bugzilla.xamarin.com/show_bug.cgi?id=52290 - monotouch-test - Unified 64bits

@rolfbjarne add one simple test in `MDLObjectTest` covering the new `Protocol` constructor.

Also fixed the `introspection-mac` tests.
